File: 백준/Gold/18114. 블랙 프라이데이/블랙 프라이데이.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int N, C;
    cin >> N >> C;
    
    vector<int> w(N);
    for (int i = 0; i < N; i++) {
        cin >> w[i];
        
        if (w[i] == C) {
            cout << 1 << endl;
            return 0;
        }
    }
    
    sort(w.begin(), w.end());
    int left = 0;
    int right = N - 1;
    int weight;
    
    while (left < right) {
        weight = w[left] + w[right];
        if (weight > C) {
            right--;
        } else if (weight == C) {
            cout << 1 << endl;
            return 0;
        } else {
            if (find(w.begin() + left + 1, w.begin() + right, C - weight) != w.begin() + right) {
                cout << 1 << endl;
                return 0;
            }
            left++;
        }
    }
    
    cout << 0 << endl;
    return 0;
}

File: 백준/Gold/18114. 블랙 프라이데이/블랙 프라이데이.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());		// 물건의 개수
		int C = Integer.parseInt(st.nextToken());		// 무게

		st = new StringTokenizer(br.readLine());
		ArrayList<Integer> w = new ArrayList<>();
		for(int i = 0; i < N; i++) {
			w.add(Integer.parseInt(st.nextToken()));
			// 1개 선택
			if(w.get(i) == C) {
				System.out.println(1);
				return;
			}
		}
		
		Collections.sort(w);
		int left = 0;
		int right = N-1;
		int weight;
		while(left < right) {
			weight = w.get(left) + w.get(right);
			if(weight > C) {
				right--;
			}
			else if(weight == C) {
				System.out.println(1);
				return;
			}
			else {
				if(w.indexOf(C-weight) < right && w.indexOf(C-weight) > left){
					System.out.println(1);
					return;
				}
				left++;
			}
		}
		System.out.println(0);
	}	
}

